core/math/basis.cpp

```diff
@@ -815,29 +815,29 @@ void Basis::get_axis_angle(Vector3 &r_axis, real_t &r_angle) const {
 #ifdef MATH_CHECKS
 	ERR_FAIL_COND(!is_rotation());
 #endif
-*/
-	real_t angle, x, y, z; // variables for result
-	real_t angle_epsilon = 0.1; // margin to distinguish between 0 and 180 degrees
-
-	if ((Math::abs(rows[1][0] - rows[0][1]) < CMP_EPSILON) && (Math::abs(rows[2][0] - rows[0][2]) < CMP_EPSILON) && (Math::abs(rows[2][1] - rows[1][2]) < CMP_EPSILON)) {
-		// singularity found
-		// first check for identity matrix which must have +1 for all terms
-		// in leading diagonal and zero in other terms
-		if ((Math::abs(rows[1][0] + rows[0][1]) < angle_epsilon) && (Math::abs(rows[2][0] + rows[0][2]) < angle_epsilon) && (Math::abs(rows[2][1] + rows[1][2]) < angle_epsilon) && (Math::abs(rows[0][0] + rows[1][1] + rows[2][2] - 3) < angle_epsilon)) {
-			// this singularity is identity matrix so angle = 0
+	*/
+
+	// http://www.euclideanspace.com/maths/geometry/rotations/conversions/matrixToAngle/index.htm
+	real_t x, y, z; // variables for result
+
+	if ((Math::abs(rows[0][1] - rows[1][0]) < CMP_EPSILON) && (Math::abs(rows[0][2] - rows[2][0]) < CMP_EPSILON) && (Math::abs(rows[1][2] - rows[2][1]) < CMP_EPSILON)) {
+		// Singularity found.
+		// First check for identity matrix which must have +1 for all terms in leading diagonal and zero in other terms.
+		if (is_diagonal() && (Math::abs(rows[0][0] + rows[1][1] + rows[2][2] - 3) < 3 * CMP_EPSILON)) {
+			// This singularity is identity matrix so angle = 0.
 			r_axis = Vector3(0, 1, 0);
 			r_angle = 0;
 			return;
 		}
-		// otherwise this singularity is angle = 180
-		angle = Math_PI;
+		// Otherwise this singularity is angle = 180.
 		real_t xx = (rows[0][0] + 1) / 2;
 		real_t yy = (rows[1][1] + 1) / 2;
 		real_t zz = (rows[2][2] + 1) / 2;
-		real_t xy = (rows[1][0] + rows[0][1]) / 4;
-		real_t xz = (rows[2][0] + rows[0][2]) / 4;
-		real_t yz = (rows[2][1] + rows[1][2]) / 4;
-		if ((xx > yy) && (xx > zz)) { // rows[0][0] is the largest diagonal term
+		real_t xy = (rows[0][1] + rows[1][0]) / 4;
+		real_t xz = (rows[0][2] + rows[2][0]) / 4;
+		real_t yz = (rows[1][2] + rows[2][1]) / 4;
+
+		if ((xx > yy) && (xx > zz)) { // Rows[0][0] is the largest diagonal term.
```

Comment:
We have a rule to start comments with a capital letter, but here it's a variable identifier so it should be kept as in code (same below).
```suggestion
		if ((xx > yy) && (xx > zz)) { // rows[0][0] is the largest diagonal term.
```